Use shared configuration for shortcode- and block-based checkout flows

reykjalin opened the pull request:
# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #1516 

1. Refactors the JS configuration preparation into static functions.
    - I'm a little unsure of this approach; is there a better way to make these configurations available to the Blocks Support class?
2. Calls the static JS configuration functions and uses the result for the blocks configuration.
3. Make sure only the Blocks JS is loaded on blocks pages.


# Testing instructions

Rather convoluted; test all the general payment flows, e.g.

- Payment with new card, saved card (Blocks + Shortcode)
- Payment with Payment Request Buttons (PRBs) (Blocks + Shortcode)
- 3DS with both PRBs, new card, saved cards (Blocks + Shortcode)
- Subscription payments (Blocks + Shortcode + Subscriptions >= 3.1.0)
    - Including renewals.
- Pre-Orders (Shortcode only, Blocks are not supported)
- Rendering of default, branded, and custom PRBs.
- Failed payments (Blocks + Shortcode)

Things to watch out for:

- Any sort of JS errors.
- Configurations that don't make it to the front-end (probably associated with JS errors).
- Discrepancies between the Shortcode- and Blocks-based cart/checkout pages.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
move javascript config generation to separate function

Commit:
move PRB JS configuration to static function

Commit:
move gateway JS configuration to static function

Commit:
Fix undefined function issue when editing pages

Commit:
use the static functions to get JS configuration for blocks

Commit:
rename javascript_configuration_object to javascript_configuration

Commit:
fix comment

Commit:
fix incorrect use of `inline_cc_form`

Commit:
remove unused functions

Commit:
Only load block JS on pages with cart/checkout block

Commit:
remove usage of old global constants in JS

Commit:
Revert "use the static functions to get JS configuration for blocks"

This reverts commit d2bb0d07a9b96f27f587760f14c8dd6f5965109f.

Commit:
Revert "Fix undefined function issue when editing pages"

This reverts commit c8d05efd1640db9d4e946ae37420fc1eff188005.

Commit:
Revert "move gateway JS configuration to static function"

This reverts commit 6a67c67680389673c1d13dbf453c7847314805a2.

Commit:
Revert "move PRB JS configuration to static function"

This reverts commit 51d777e120ca401a3274500109db5eae0649223b.

Commit:
Move PRB js configuration object to separate function

Commit:
use PRB and gateway JS configuration with Blocks

Commit:
use PRB configuration to get PRB button label for blocks

Commit:
provide the Payment Request configuration to Blocks Support class via DI

dechov commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -361,54 +361,12 @@ public function admin_scripts() {
-			return;
-		}
-
-		$current_theme = wp_get_theme();
```
Nice catch – this variable was unused.

dechov commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -81,26 +93,55 @@ public function get_payment_method_script_handles() {
+
+		$gateways = WC()->payment_gateways->get_available_payment_gateways();
+		if ( isset( $gateways['stripe'] ) ) {
+			$js_configuration = $gateways['stripe']->javascript_configuration();
```
Should these use `$this->name` instead?

reykjalin replied:
Hmmm, I don't think so because this should be the `id` defined in the gateway, like this one:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/cc9e6e61335ba3bbe3a5924650612abc86f8ec43/includes/class-wc-gateway-stripe.php#L87

And trying to use that becomes sort of circular doesn't it? Instantiating a second gateway to get the ID, only to then use that ID to check for the existence of another gateway?

I agree with your sentiment though; using a string there seems fairly fragile. Maybe we should add a constant with the gateway name somewhere and use that as the ID for the gateway?

Just to be clear; `$this->name` is used to identify the payment method to the Blocks plugin I believe; it's not connected to the Payment Gateway code AFAICT 🤔

dechov replied:
Ah, I assumed that the `name` was already bound to be identical to the payment gateway ID for it to work, but I'm not sure if that's true – we can leave this alone 👍

reykjalin replied:
Yeah, we should use the same value for our own sanity, but it's not actually a requirement.

I'll mark this as resolved, but feel free to open the discussion on this again 👍

dechov commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -361,54 +361,12 @@ public function admin_scripts() {
+	 * @return array  The configuration object to be loaded to JS.
 	 */
-	public function payment_scripts() {
+	public function javascript_configuration() {
```
Just a slight preference on my part for something like `js_config` (or `_params` perhaps) – since it's more concise without losing comprehensibility IMO – but no need to change if you'd intentionally named it this way instead.

reykjalin replied:
I do prefer to be explicit, which is why I went with the longer name. It's not really that cumbersome to write when you have autocomplete anyway. I'd like to at least keep the `javascript_` part, because `js` is relatively ambiguous in this context IMO.

But on that note, which do you think is the best descriptor for this; `javascript_parameters`/`javascript_params` or `javascript_configuration`/`javascript_config`?

dechov replied:
Hm, not a huge deal but I'd vote for `javascript_params` since "configuration" sort of implies to me that it's just user settings when in fact there are also return URLs, nonces, etc., just everything needed for parameterizing the client – what do you think?

reykjalin replied:
Your reasoning makes sense, and `javascript_params` sounds good to me 🙂 

Changed the names in c36fa93b5ca73332d2b2400cb2ee4f8abe78db9b 🙌

dechov commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -468,7 +426,66 @@ public function payment_scripts() {
+		}
+
+		// If we're loading the cart or checkout block the scripts here should not be loaded.
+		if ( WC_Stripe_Helper::has_cart_or_checkout_block_on_current_page() ) {
```
As I understand, the idea here is to avoid loading the (shortcode-specific) script when the _shortcode isn't present_, not when a _block is present_. There may be little practical difference (except in extreme edge cases like the cart block sharing a page with checkout shortcode), but looking at the `is_cart` condition [here](https://github.com/woocommerce/woocommerce/blob/0507f06508a448995ae6ebec903115e670cc5ee4/includes/wc-conditional-functions.php#L98), it looks like there's a handy [`wc_post_content_has_shortcode` function](https://github.com/woocommerce/woocommerce/blob/0507f06508a448995ae6ebec903115e670cc5ee4/includes/wc-conditional-functions.php#L412-L422) we can use here.

reykjalin replied:
This is fantastic observation, it really didn't occur to me 🥇 

Made the suggested change in 34935a0f8a8060efa498baa730325167ec2e5411. Awesome catch! 👏 🙌

dechov replied:
Nice – ends up more being concise than I'd expected :+1:

dechov reviewed (changes requested):
Nice improvement! The refactor looks good in general, and I've verified that the "stripe.js" front-end script asset loads only on _shortcode_ cart/checkout, not for the block :+1:

I was originally seeing an error in the Checkout block due to `publicKey` being passed, since we do actually still want the (reverted) change [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1585/commits/6f2f779df93acba93f443d25e73620de6542edd4#diff-dd25ac78a39d9f65aa3b146adc56da769e8cacea23a97d1921b334c17418f563L39-R39).

With that change, payments (incl. 3DS) work with new and saved cards, with both shortcode and blocks. However, Apple Pay in the Cart and Checkout blocks is failing for me, with this error:
> `Unhandled Promise Rejection: ReferenceError: Can't find variable: wc_stripe_payment_request_params`

(I can test some more cases tomorrow.)

Commit:
fix issues from a bad revert

reykjalin commented:
> I was originally seeing an error in the Checkout block due to publicKey being passed, since we do actually still want the (reverted) change here.
>
> With that change, payments (incl. 3DS) work with new and saved cards, with both shortcode and blocks. However, Apple Pay in the Cart and Checkout blocks is failing for me, with this error:
>
> ```
> Unhandled Promise Rejection: ReferenceError: Can't find variable: wc_stripe_payment_request_params
> ```
> (I can test some more cases tomorrow.)

Ugh, sorry about this 🤦 
I must've screwed the revert up while fixing the conflicts that happened during the revert. Sorry about that!

I've pushed a fix in cc9e6e61335ba3bbe3a5924650612abc86f8ec43 and ran some tests to verify that it fixes the issue. Sorry about that!

Commit:
Change how we decide when shortcode JS is loaded

Instead of not loading the shortcode JS when a Block is present on the
current page, we now load the shortcode JS when a cart or checkout
shortcode is present on the page.

dechov commented on includes/class-wc-stripe-helper.php:
```diff
@@ -597,4 +597,25 @@ public static function convert_wc_locale_to_stripe_locale( $wc_locale ) {
+	 * @since 5.3.0
+	 * @return boolean
+	 */
+	public static function has_cart_or_checkout_block_on_current_page() {
```
Should we remove this function since it's no longer being used?

reykjalin replied:
Yes, removed in ed27e97a734d59171aac0c469dfc861533fba865

Nice catch! 👏

Commit:
rename `javascript_configuration` to `javascript_params`

Commit:
remove unused function

dechov commented on client/blocks/credit-card/payment-method.js:
```diff
@@ -18,7 +18,7 @@ import { InlineCard, CardElements } from './elements';
 
 export const getStripeCreditCardIcons = () => {
-	return Object.entries( getStripeServerData().icons ).map(
+	return Object.entries( getStripeServerData()?.icons ).map(
```
If there is no data, I think like this particular line would result in an error with or without the optional chaining (and possibly a less helpful one this way)?

reykjalin replied:
Ah yeah, this should definitely be changed to

```js
return Object.entries( getStripeServerData()?.icons ?? {} ).map(
```

> I think like this particular line would result in an error with or without the optional chaining (and possibly a less helpful one this way)?

I think the error isn't helpful in any case to be honest 😅 

<img width="573" alt="image" src="https://user-images.githubusercontent.com/13835680/121271550-7904f000-c8b3-11eb-852d-4f5fe4a3211f.png">

And if there's an error there that crashes the rest of the Blocks support JS I think. I'd rather use the null coalescing and just add a `?? {}` so that the Stripe blocks support doesn't crash if we happen to forget to supply icons 😬

reykjalin replied:
Sorry; just to clarify my point with the error ^ not being useful is that when you don't see where exactly it's coming from — which we don't because webpack minifes the code during the build — that error isn't useful at all.

reykjalin replied:
Fixed in eea0676f5d0d0e6bebc7607c271c01ff74bd2d88

dechov replied:
Sounds good. And I really have no idea how that "like" snuck in to my comment there 😆

dechov commented on client/blocks/payment-request/index.js:
```diff
@@ -32,10 +30,9 @@ const paymentRequestPaymentMethod = {
-		if ( typeof wc_stripe_payment_request_params === 'undefined' ) {
+		// If the JS Payment Request Button configuration object loaded from PHP is not available
+		// we don't support payment requests.
+		if ( ! getStripeServerData()?.button ) {
```
Is there any known case where this could be false(-y)?

reykjalin replied:
There used to be before I added [the `button` prop](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1585/files#diff-3197fd215f7f836c9060fc595ce3eef6b21515846bb890f8d3e0a273b605ce28R109-R111) to the blocks code every time.

Nice catch! We can remove that check 🙂

reykjalin replied:
Removed in a62f1744add8409df2f94651ef3297b65614ece6

dechov commented on client/blocks/stripe-utils/normalize.js:
```diff
@@ -11,6 +9,11 @@
+/**
+ * Internal dependencies
+ */
+import { getStripeServerData } from '../stripe-utils';
```
A bit of an odd path, going out of `stripe-utils` and back in – maybe should be `'./utils'` like in the `load-stripe.js` file?

reykjalin replied:
Yes, I'll change that, nice catch!

reykjalin replied:
Fixed in da51b88aa60ddc7a3842e246e32d71e2190e1038

dechov commented on client/blocks/stripe-utils/type-defs.js:
```diff
@@ -280,11 +280,9 @@
+ * @property {string}                      key                  The public api key for stripe
  *                                                              requests.
- * @property {boolean}                     allowPrepaidCard     True means that prepaid cards
+ * @property {boolean}                     allow_prepaid_card   True means that prepaid cards
```
I believe this is a `string` as well.

reykjalin replied:
Fixed in 677655fb32d043ba61f6cf46bbea39eb845b0699

dechov commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -81,26 +93,55 @@ public function get_payment_method_script_handles() {
+
+		$gateways = WC()->payment_gateways->get_available_payment_gateways();
+		if ( isset( $gateways['stripe'] ) ) {
+			$js_configuration = $gateways['stripe']->javascript_params();
```
Hm, looks like the `get_supported_features` method below just instantiates a `new WC_Gateway_Stripe()`… think we should change one or the other for consistency, or leave them as they are?

reykjalin replied:
Nice catch!

Fixed in 0c9e774c82853a3cfcac3844031f59609ad981a1

dechov reviewed (commented):
Appears to work well :+1: – I'll do just a bit more testing in a bit.

Minor note: In general, are there known cases in which `getStripeServerData()` will be null, and all the optional chaining would make a difference? Looks like it was added to many of the calls but not all – is there a rule here?

reykjalin commented:
> In general, are there known cases in which getStripeServerData() will be null, and all the optional chaining would make a difference? Looks like it was added to many of the calls but not all – is there a rule here?

Not any _known_ cases; at least not in terms of a successful flow. That said, Darren mentioned in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1467#discussion_r602778824 on the original Blocks support PR that this function could potentially return `null`, and there's no real guarantee that the shape of the JS object doesn't change down the line.

So I guess you're mostly wondering if it'd be better to have the front-end crash if we change something during development rather than handle the error gracefully without any notification that's what happened?
I'm not sure what's better, but I personally lean towards handling this silently with null coalescing. What do you think?

Commit:
fix null coalescing issue

Commit:
remove unnecessary check when deciding whether to display PRBs

Commit:
improve import path

Commit:
fix incorrect typedef

Commit:
use WC()->payment_gateways global to retrieve Stripe gateway

dechov commented:
Well I meant in terms of a _possible_ flow, not necessarily successful. I'm fine with the optional chaining but was mostly wondering why it's in some places and not others, and if there's a pattern there – this example (among others) was left alone: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0c9e774c82853a3cfcac3844031f59609ad981a1/client/blocks/credit-card/use-payment-processing.js#L63

Commit:
add missing optional chaining operators

reykjalin commented:
Thanks for calling out the lack of optional chaining Paul!

I've fixed all instances I could find in df49b46d8214f4f0f41ef54f781291529a925c41 🙂

Commit:
add changelog entries

dechov reviewed (approved):
All LGTM 🎉 – tested most of the cases in the testing instructions (though not every single permutation) and some others, and tried various config like PRB settings, inline form, disallowing prepaid cards.
